src/sbuf/SBuf.h

```diff
@@ -580,28 +451,50 @@ class SBuf
     /// std::string export function
     std::string toStdString() const { return std::string(buf(),length()); }
 
-    const_iterator begin() const {
-        return const_iterator(*this, 0);
-    }
-
-    const_iterator end() const {
-        return const_iterator(*this, length());
-    }
-
-    const_reverse_iterator rbegin() const {
-        return const_reverse_iterator(*this, length());
-    }
+    /// SBuf object identifier meant for test cases and debugging.
+    /// Does not change when object does, including during assignment.
+    const InstanceId<SBuf> id;
 
-    const_reverse_iterator rend() const {
-        return const_reverse_iterator(*this, 0);
+    /* std::string API */
+    SBuf& operator =(const SBuf &S) { return assign(S); } ///< \sa SBuf::assign
+    SBuf& operator =(SBuf &&S) /**< \sa SBuf::assign */ {
+        ++stats.moves;
+        if (this != &S) {
+            store_ = std::move(S.store_);
+            off_ = S.off_;
+            len_ = S.len_;
+            S.store_ = nullptr; //RefCount supports NULL, and S is about to be destructed
+            S.off_ = 0;
+            S.len_ = 0;
+        }
+        return *this;
     }
-
+    SBuf& operator =(const char *S) {return assign(S);} ///< \sa SBuf::assign
+    const_iterator begin() const { return const_iterator(*this, 0); }
+    const_iterator end() const { return const_iterator(*this, length()); }
+    const_reverse_iterator rbegin() const { return const_reverse_iterator(*this, length()); }
+    const_reverse_iterator rend() const { return const_reverse_iterator(*this, 0); }
+    // size_type size() const noexcept { return length(); } // XXX: prohibited by QA reviewer 2014
+    size_type length() const noexcept { return len_; } ///< \returns the number of bytes stored in SBuf
+    // TODO: size_t max_size() const noexcept { return SBuf::maxSize; }
+    void clear() noexcept;
+    // TODO: bool empty() const noexcept { return isEmpty(); }
```

Comment:
This is part of the STL defined API, it may be required for some container algorithms.
https://cplusplus.com/reference/stl/

Reply:
Yes, an empty() method is a C++ [Container](https://en.cppreference.com/w/cpp/named_req/Container) requirement.

If we want SBuf to satisfy C++ Container requirement, let's document _that_ as a TODO (instead of adding a TODO for every currently missing Container method as this PR starts to do). Adding such a general TODO would also avoid discussing (in this PR) whether to eliminate isEmpty() in the process (a decision we would have to make at some point but do not have to make here and now). And we would not need to discuss problems with the suggested implementation of those future Container methods!

Satisfying Container requirement is a pretty tall order though -- SBuf is probably missing a lot of the required methods (and might even violate some of the required APIs; I have not checked).